lib/pbio/include/pbsys/main.h

```diff
@@ -11,9 +11,36 @@
 
 #include <pbsys/config.h>
 
+#include <stdbool.h>
+#include <stdint.h>
+
+/**
+ * Main application program data information.
+ */
+typedef struct _pbsys_main_program_t {
+    /**
+     * User program data.
+     */
+    uint8_t *data;
+    /**
+     * User program data size.
+     */
+    uint32_t size;
+    /**
+     * Whether to run an application-specific builtin program instead of the
+     * program given by the data. The builtin program may still use the data.
+     */
+    bool run_builtin;
+} pbsys_main_program_t;
+
 #if PBSYS_CONFIG_MAIN
 
-void pbsys_main_application(void);
+void pbsys_main_application(pbsys_main_program_t *program);
+
+#else
+
+static inline void pbsys_main_application(pbsys_main_program_t *program) {
```

Comment:
Wouldn't this just cause an infinite loop? Maybe better to just get a compiler error if the symbol is not defined in this case.